### PR TITLE
Fix Inductor replacement patterns across view-to-reshape

### LINKED ISSUE #151649: Inductor pattern matcher replaces aten.reshape with aten.view in pattern
### 🐛 Describe the bug

Based on my rudimentary debugging, it seems that the pattern will record a `torch.ops.aten.reshape.default` as a `aten.view.default`. When a reshape is found in the fx.Graph, it is hence not matched.

In the example below, we compile the pattern itself, which should obviously match, but it doesn't.

```python
from typing import Callable, List

import torch
from torch._inductor.pattern_matcher import register_replacement, PatternMatcherPass, fwd_only


def pattern(x, y):
    c = torch.mm(x, y)
    d = torch.ops.aten.reshape.default(c, [-1, 4, 4])
    return d.relu()

def replacement(x, y):
    c = torch.mm(x, y)
    return torch.ops.aten.reshape.default(c.relu(), [-1, 4, 4])


inputs = [
    torch.empty(5, 16, device="cuda"),
    torch.empty(16, 16, device="cuda"),
]

patterns = PatternMatcherPass()
register_replacement(pattern, replacement, inputs, fwd_only, patterns)

def pg_pass(graph: torch.fx.Graph):
    count = patterns.apply(graph)
    print(f"Count: {count}")

def custom_backend(graph: torch.fx.GraphModule, example_inputs: List[torch.Tensor]) -> Callable:
    from torch._inductor import config
    current_config = config.shallow_copy_dict()
    from torch._inductor.compile_fx import compile_fx
    current_config['post_grad_custom_post_pass'] = pg_pass
    return compile_fx(graph, example_inputs, config_patches=current_config)


func = torch.compile(pattern, backend=custom_backend)

# The line below did not work somehow?
# func = torch.compile(pattern, backend="inductor", options={"post_grad_custom_post_pass":pg_pass})

func(*inputs)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.6.0+rocm6.2.4
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.2.41134-65d174c3e

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 10.5.0-1ubuntu1~22.04) 10.5.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-131-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Instinct MI300X (gfx942:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: 6.2.41134
MIOpen runtime version: 3.2.0
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               104
On-line CPU(s) list:                  0-103
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8470
CPU family:                           6
Model:                                143
Thread(s) per core:                   1
Core(s) per socket:                   52
Socket(s):                            2
Stepping:                             8
BogoMIPS:                             4000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
L1d cache:                            4.9 MiB (104 instances)
L1i cache:                            3.3 MiB (104 instances)
L2 cache:                             208 MiB (104 instances)
L3 cache:                             210 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-51
NUMA node1 CPU(s):                    52-103
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] pytorch-triton-rocm==3.2.0
[pip3] torch==2.6.0+rocm6.2.4
[pip3] torchaudio==2.6.0+rocm6.2.4
[pip3] torchvision==0.21.0+rocm6.2.4
[pip3] triton==3.2.0
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #184109

Allow register_replacement patterns traced through view to match post-grad graphs rewritten to reshape, while preserving exact matching for manual patterns.

Fixes #151649

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — COMMENTED** [review]
Hmm, when do we even insert the reshapes ? I remember it had something to do with channels last conversion. Ideally the post grad IR is supposed to be exact aliasing. 

Can you investigate why we insert them in the first place ? An alternative would be to use clone + view as needed. Assuming we're able to remove the clones. 

And then, make the pattern matcher somehow handle it.

**Comment by jansel** [comment]
My agent says I rebased this PR on current main and resolved the conflict in pattern_matcher.py by preserving main's structured kwargs restoration plus the replacement-graph reshape normalization. I investigated the reshape source: compile_fx runs view_to_reshape before fake tensor propagation because channels-last layout optimization can make a previously valid view illegal; clone+view would be a broader IR invariant/copy-placement change, so this PR keeps the fix scoped to register_replacement by matching view/reshape equivalently only for ReplacementPatternEntry while manual patterns remain exact.

**Review by zou3519 — COMMENTED** [review]
I think we should avoid converting view to reshape. There should be some invariant that all ops in the inductor graph are just post-grad aten, and reshape is not post-grad-aten.

**Comment by jansel** [comment]
My agent says I investigated this again. I agree that a stricter post-grad/core-aten invariant would be cleaner, but current Inductor intentionally runs view_to_reshape before FakeTensorProp and before post_grad_custom_post_pass because layout optimization can make a previously valid view illegal after channels-last conversion. That behavior is documented in compile_fx.py next to the call, is recursive through subgraphs, and reshape is already handled by Inductor lowering. This PR does not introduce reshape into the post-grad IR; it keeps the fix scoped to register_replacement so generated replacement patterns can match the IR that custom post-grad passes already receive, while manual view patterns remain exact. Replacing view_to_reshape with clone+view would be a broader layout/copy-placement invariant change, so I think it should be handled separately from this bug fix.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and resolved the pattern_matcher conflict while preserving current structured-match and mutation-region handling. Fresh review also found and fixed a stride-safety gap: equivalence matches now retrace precompiled patterns and reject non-viewable reshapes, while nested replacement subgraphs are normalized and recompiled. The focused matcher/mutation tests, serialized-pattern check, end-to-end issue reproducer, and lint all pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
